include/envoy/server/filter_config.h

```diff
@@ -296,14 +296,12 @@ class NamedHttpFilterConfigFactory {
   virtual ProtobufTypes::MessagePtr createEmptyConfigProto() { return nullptr; }
 
   /**
-   * @return ProtobufTypes::MessagePtr create an empty virtual host, route, or weight cluster-local
-   *         config proto message for v2. The filter config, which arrives in an opaque
-   *         google.protobuf.Struct message, will be converted to JSON and then parsed into this
-   *         empty proto. By default, this method returns the same value as createEmptyConfigProto,
-   *         and can be optionally overridden in implementations.
+   * @return RouteSpecificFilterConfigConstSharedPtr allow the filter to pre-process per route
+   * config. Returned object will be stored in the loaded route configuration.
    */
-  virtual ProtobufTypes::MessagePtr createEmptyRouteConfigProto() {
-    return createEmptyConfigProto();
+  virtual Router::RouteSpecificFilterConfigConstSharedPtr
+  createRouteSpecificFilterConfig(const ProtobufWkt::Struct&) {
```

Comment:
I would strongly -1 to exposing Struct to filter interface and in favor of wrapping them in an interface like https://github.com/envoyproxy/envoy/pull/3109 did. Hiding Struct from filter interface is important when we see scaling issues with the config and want to migrate to Any in future, as mentioned in https://github.com/envoyproxy/data-plane-api/issues/540#issuecomment-373250850 . cc @htuch.

Reply:
+1000